File: 151004/Tishalovich/rest_basics/src/main/java/by/bsuir/dc/rest_basics/services/impl/MessageServiceImpl.java
package by.bsuir.dc.rest_basics.services.impl;

import by.bsuir.dc.rest_basics.dal.MessageDao;
import by.bsuir.dc.rest_basics.entities.Message;
import by.bsuir.dc.rest_basics.entities.dtos.request.MessageRequestTo;
import by.bsuir.dc.rest_basics.entities.dtos.response.MessageResponseTo;
import by.bsuir.dc.rest_basics.services.MessageService;
import by.bsuir.dc.rest_basics.services.exceptions.ApiException;
import by.bsuir.dc.rest_basics.services.exceptions.GeneralSubCode;
import by.bsuir.dc.rest_basics.services.exceptions.MessageSubCode;
import by.bsuir.dc.rest_basics.services.impl.mappers.MessageMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service
@RequiredArgsConstructor
public class MessageServiceImpl implements MessageService {

    private final MessageDao dao;

    private final MessageMapper mapper;

    @Override
    public MessageResponseTo create(MessageRequestTo requestTo) throws ApiException {
        Message model = mapper.requestToModel(requestTo);
        Message savingRes;
        try {
            savingRes = dao.save(model);
        } catch (DataIntegrityViolationException e) {
            throw new ApiException(
                    HttpStatus.FORBIDDEN.value(),
                    MessageSubCode.CONSTRAINT_VIOLATION.getSubCode(),
                    e.getMessage()
            );
        }
        return mapper.modelToResponse(savingRes);
    }

    @Override
    public List<MessageResponseTo> getAll() {
        Iterable<Message> models = dao.findAll();
        return StreamSupport
                .stream(models.spliterator(), false)
                .map(mapper::modelToResponse)
                .collect(Collectors.toList());
    }

    @Override
    public MessageResponseTo get(Long id) throws ApiException {
        Optional<Message> author = dao.findById(id);

        if (author.isEmpty()) {
            throw new ApiException(
                    HttpStatus.NOT_FOUND.value(),
                    GeneralSubCode.WRONG_ID.getSubCode(),
                    GeneralSubCode.WRONG_ID.getMessage()
            );
        }

        return mapper.modelToResponse(author.get());
    }

    @Override
    public MessageResponseTo update(MessageRequestTo requestTo) throws ApiException {
        Message model = mapper.requestToModel(requestTo);

        Message updateRes = dao.save(model);
        //if (updateRes == null) {
        //    throw new ApiException(
        //            HttpStatus.NOT_FOUND.value(),
        //            GeneralSubCode.WRONG_ID.getSubCode(),
        //            GeneralSubCode.WRONG_ID.getMessage()
        //    );
        //}

        return mapper.modelToResponse(updateRes);
    }

    @Override
    public void delete(Long id) throws ApiException {
        if (!dao.existsById(id)) {
            throw new ApiException(
                    HttpStatus.NOT_FOUND.value(),
                    GeneralSubCode.WRONG_ID.getSubCode(),
                    GeneralSubCode.WRONG_ID.getMessage()
            );
        }

        dao.deleteById(id);
    }

}

File: 151004/Burbouski/appLab1/src/main/kotlin/com/github/hummel/dc/lab1/service/impl/MessageServiceImpl.kt
package com.github.hummel.dc.lab1.service.impl

import com.github.hummel.dc.lab1.bean.Message
import com.github.hummel.dc.lab1.dto.request.MessageRequestTo
import com.github.hummel.dc.lab1.dto.request.MessageRequestToId
import com.github.hummel.dc.lab1.dto.response.MessageResponseTo
import com.github.hummel.dc.lab1.service.MessageService
import com.github.hummel.dc.lab1.util.BaseRepository

class MessageServiceImpl(
	private val messageRepository: BaseRepository<Message, Long>
) : MessageService {

	override fun getAll(): List<MessageResponseTo> {
		val result = messageRepository.data.map { it.second }

		return result.map { it.toResponse() }
	}

	override fun create(messageRequestTo: MessageRequestTo?): MessageResponseTo? {
		val id = if (messageRepository.data.isEmpty()) {
			-1
		} else {
			messageRepository.getLastItem()?.id ?: return null
		} + 1

		val bean = messageRequestTo?.toBean(id) ?: return null
		val result = messageRepository.addItem(bean.id, bean)

		return result?.toResponse()
	}

	override fun deleteById(id: Long): Boolean = messageRepository.removeItem(id)

	override fun getById(id: Long): MessageResponseTo? {
		val result = messageRepository.getItemById(id)?.second

		return result?.toResponse()
	}

	override fun update(messageRequestToId: MessageRequestToId?): MessageResponseTo? {
		val bean = messageRequestToId?.toBean() ?: return null
		val result = messageRepository.addItem(bean.id, bean)

		return result?.toResponse()
	}
}